### ISSUE TITLE
How to tuning the load model's prediction

### ISSUE THREAD

**Issue by weidong8405347** [Microsoft/LightGBM#1123]
i want to load a model as base model, but the base model not suit for the new data, so i want to tuning the base model's leaf prediction, i use the code like this, it failed, because the datapartition can not recognize the feature value, so i wonder if  have a easy way to deal with these problem

void GBDT::TuneTree() {
  if (objective_function_ == nullptr) {
    Log::Fatal("No object function provided");
  }         
  Log::Info("tune tree start, size: %d", models_.size());
  score_updater_ =  std::unique_ptr<ScoreUpdater>(new ScoreUpdater(train_data_, 1));
  score_updater_->ResetZero();
  for (size_t index = 0; index < models_.size(); ++index) {
      objective_function_->
        GetGradients(online_score_updater_->score(), gradients_.data(), hessians_.data());
      auto tree = tree_learner_->FitByExistingTree(models_[index].get(), gradients_.data(), hessians_.data());
      models_[index].reset(tree);
      score_updater_->AddScore(tree_learner_.get(), models_[index].get(), 0);
  }         
  train_score_updater_->CopyScore(score_updater_->score());
}

**Comment by guolinke** [comment]
you should use this function to ```AddScore``` : https://github.com/Microsoft/LightGBM/blob/master/src/boosting/score_updater.hpp#L75-L77

**Comment by weidong8405347** [comment]
@guolinke  i already use the function, but it does not work, i woder that the model feature not match the data partition, you can see the code
becase the tree is load from input model not split by tree learner

**Comment by weidong8405347** [comment]
@guolinke   FitByExistingTree function is not work too
because data_partition_->GetIndexOnLeaf(i, &cnt_leaf_data) can not get cnt_leaf_data

**Comment by guolinke** [comment]
@weidong8405347 please try this PR.

**Comment by weidong8405347** [comment]
@guolinke  thanks, i have one question is that your code is the demo? because i did not see the implement of gbdt.cpp RefitTree

**Comment by guolinke** [comment]
@weidong8405347 
https://github.com/Microsoft/LightGBM/pull/1124/files#diff-f9317e8f1747aa12b0ba34988880e97c

**Comment by weidong8405347** [comment]
@guolinke  Thank you

**Comment by guolinke** [comment]
@weidong8405347 you can provide the feedback if met any problems.

**Comment by weidong8405347** [comment]
@guolinke  thank you so much, but i still have one more need, the leaf_pred should base the predictor to predict, but the function is too heavy, when i train the tree, i want to fine tune the other tree, but i can not get leaf_pred ,can you help me? the code is below:
when i finish train the last tree, i want to fine tune the other tree,
for example , when i trained the Nth tree, i want to fine tune the N-1th tree and N-2the tree, so when i fine tune the N-1th tree, i need get N-1th leaf_pred to fitby, when i fine tune the N-2the tree, i need get N-2th leaf_pred to fitby
bool GBDT::TrainOneIter {
      // some orgin code
      // add some new code
       for (size_t i = 0; i < corrective_num; ++i, ++model_start_index) {
           size_t model_index = model_start_index % model_size + num_init_iteration_;
           models_[model_index]->Shrinkage(-1.0);
           train_score_updater_->AddScore(models_[model_index].get(), cur_tree_id);
           for (auto& score_updater : valid_score_updater_) {
               score_updater->AddScore(models_[model_index].get(), cur_tree_id);
           }   
           models_[model_index]->Shrinkage(-1.0);
                
           auto tree = tree_learner_->FitByExistingTree(models_[model_index].get(), gradients_.data(), hessians_.data());
           train_score_updater_->AddScore(tree_learner_.get(), tree, cur_tree_id);
           for (auto& score_updater : valid_score_updater_) {
               score_updater->AddScore(tree_learner_.get(), models_[model_index].get(), cur_tree_id);
           }   
           models_[model_index].reset(tree);
        }
}

**Comment by guolinke** [comment]
@weidong8405347 
two solutions:
1. snapshot the ```data_partition_``` (https://github.com/Microsoft/LightGBM/blob/master/src/treelearner/serial_tree_learner.h#L112) for the iterations(trees) you need to fine tune. Then when you need to fine tune, you can use the according ```data_partition``` in tree_learner, so that your above code can work.
2. If you are using other dataset to fine tune (not the training data), the solution 1 cannot work. But you can implement a new pred-leaf function based on ```Dataset``` object, which is quite easy. However, please note that the dataset used for fine tune should be loaded as validation data, or the result will be wrong.

**Comment by weidong8405347** [comment]
@guolinke  thanks for the advice,
the first advice means i should maintain a vector to store the every tree's partition? and add a new function, when FitByExistingTree use the store data partition to store? because the number of fine tune trees is large, and store the partition is too large?
the second advice dataset add a pred-leaf function, can you show me the code? to you mayby easy, but i am not familiar with the dataset class

**Comment by weidong8405347** [comment]
@guolinke  for the first advice i implement one version, but the result is not correct, can you  help me to find out where is wrong
i just to store the leaf map data index for each tree
Tree* SerialTreeLearner::FitByExistingTree(const Tree* old_tree, const score_t* gradients, const score_t *hessians, std::unordered_map<int, std::vector<data_size_t> >* leaf_map_idxs) const {
    auto tree = std::unique_ptr<Tree>(new Tree(*old_tree));
    CHECK(data_partition_->num_leaves() >= tree->num_leaves());
    OMP_INIT_EX();
    //#pragma omp parallel for schedule(static)
    for (int i = 0; i < tree->num_leaves(); ++i) {
      OMP_LOOP_EX_BEGIN();
      data_size_t cnt_leaf_data = 0;
      auto tmp_idx = data_partition_->GetIndexOnLeaf(i, &cnt_leaf_data);
      double sum_grad = 0.0f;
      //double sum_hess = 0.0f;
      double sum_hess = kEpsilon;
      //#pragma omp atomic write
      (*leaf_map_idxs)[i] = std::vector<data_size_t>(cnt_leaf_data, 0); 
      for (data_size_t j = 0; j < cnt_leaf_data; ++j) {
        auto idx = tmp_idx[j];
        sum_grad += gradients[idx];
        sum_hess += hessians[idx];
        (*leaf_map_idxs)[i][j] = idx;
      }
      // avoid zero hessians.
      double output = FeatureHistogram::CalculateSplittedLeafOutput(sum_grad, sum_hess,
                                                                    tree_config_->lambda_l1, tree_config_->lambda_l2);
      tree->SetLeafOutput(i, output * tree->shrinkage());                                                                                                               
      OMP_LOOP_EX_END();
    }  
    OMP_THROW_EX();
    return tree.release();
}
Tree* SerialTreeLearner::FitByExistingTree(const Tree* old_tree,
                                         std::unordered_map<int, std::vector<data_size_t> >& leaf_map_idxs,
                                         const score_t* gradients, const score_t *hessians) const {
    auto tree = std::unique_ptr<Tree>(new Tree(*old_tree));
    OMP_INIT_EX();
    #pragma omp parallel for schedule(static)
    for (int i = 0; i < tree->num_leaves(); ++i) {
      OMP_LOOP_EX_BEGIN();
      double sum_grad = 0.0f;
      //double sum_hess = 0.0f;
      double sum_hess = kEpsilon;
        for (auto& idx : leaf_map_idxs[i]) {
        sum_grad += gradients[idx];
        sum_hess += hessians[idx];
      }   
      // avoid zero hessians.
      //if (sum_hess <= 0) sum_hess = kEpsilon;                                                                                                                         
      double output = FeatureHistogram::CalculateSplittedLeafOutput(sum_grad, sum_hess,
                                                                    tree_config_->lambda_l1, tree_config_->lambda_l2);
      tree->SetLeafOutput(i, output * tree->shrinkage());
      OMP_LOOP_EX_END();
    }   
    OMP_THROW_EX();
    return tree.release();
}
           Tree* tree = nullptr;
           if (model_index == model_size - 1) {
               std::unordered_map<int, std::vector<data_size_t> > leaf_map_idxs;
               tree = tree_learner_->FitByExistingTree(models_[model_index].get(), gradients_.data(), hessians_.data(), &leaf_map_idxs);
               leaf_map_idxs_vec_[model_index] = std::move(leaf_map_idxs);
           } else {
               tree = tree_learner_->FitByExistingTree(models_[model_index].get(), leaf_map_idxs_vec_[model_index],
                                                       gradients_.data(), hessians_.data());
           }   

 bool GBDT::TrainOneIter {
      // some orgin code
      // add some new code                                                                                                                                         
           models_[model_index]->Shrinkage(-1.0);
           train_score_updater_->AddScore(models_[model_index].get(), cur_tree_id, leaf_map_idxs_vec_[model_index]);
           for (auto& score_updater : valid_score_updater_) {
               score_updater->AddScore(models_[model_index].get(), cur_tree_id);
           }   
           models_[model_index]->Shrinkage(-1.0);
               
           ////update new corrective tree
           train_score_updater_->AddScore(tree, cur_tree_id, leaf_map_idxs_vec_[model_index]);
           for (auto& score_updater : valid_score_updater_) {
               score_updater->AddScore(tree, cur_tree_id);
           }   
           models_[model_index].reset(tree);
        }

**Comment by weidong8405347** [comment]
@guolinke  train score updater is right, but the valid score updater not right
can you give some advice

**Comment by guolinke** [comment]
Sorry, your code is very messy. And I don't have time to find out the bug. 

for the second solution. 

refer to this function: https://github.com/Microsoft/LightGBM/blob/master/src/io/tree.cpp#L100-L157
You can add a new ```PredictionFun``` that return the leaf_index, not add the score. 

BTW, I think your code should be in ```Train``` not the ```TrainOneIter```.

**Comment by weidong8405347** [comment]
@guolinke  thanks for advice, the code i already delete
for the first solution, i do not know how to deal with the valid_score_updater
for the train_score_updater i can use data partition

**Comment by guolinke** [comment]
@weidong8405347 you don't need the data partition for validation datasets. as long as the tree is new-trained(not loaded from model file), you can use the tree to update the validation scores.
